Crypto/CertificateTest.cc

```diff
@@ -194,12 +197,168 @@ TEST_CASE("Persistent key and cert", "[Certs]") {
+    
+    // Save cert2 to an existing ID:
+    ExpectException(error::LiteCore, error::CryptoError, [&]{
+        cert2->save("cert1", true);
+    });
+    
+    // Save cert2:
+    cert2->save("cert2", true);
+    Retained<Cert> cert2a = Cert::loadCert("cert2");
+    REQUIRE(cert2a);
+    CHECK(cert2a->data() == cert2->data());
+    
+    // Delete cert1:
+    Cert::deleteCert("cert1");
+    CHECK(Cert::loadCert("cert1") == nullptr);
+    
+    // Delete cert2:
+    Cert::deleteCert("cert2");
+    CHECK(Cert::loadCert("cert2") == nullptr);
+}
+
+
+TEST_CASE("Persistent cert chain", "[Certs]") {
+    // Create a CA Cert:
+    Retained<PrivateKey> caKey = PrivateKey::generateTemporaryRSA(2048);
+    Cert::IssuerParameters caIssuerParams;
+    caIssuerParams.is_ca = true;
+    Retained<Cert> caCert = new Cert(DistinguishedName(kCAName), caIssuerParams, caKey);
+    cerr << "CA cert info:\n" << string(caCert->summary("\t"));
+    
+    // Create CSR1:
+    Retained<PrivateKey> key1 = PrivateKey::generateTemporaryRSA(2048);
+    Retained<CertSigningRequest> csr1 = new CertSigningRequest(DistinguishedName(kSubjectName), key1);
+    CHECK(csr1->subjectName() == kSubjectName);
+    CHECK(csr1->subjectPublicKey()->data(KeyFormat::Raw) == key1->publicKey()->data(KeyFormat::Raw));
+    
+    // Sign and create cert1 with the CA Cert:
+    Cert::IssuerParameters caClientParams1;
+    caClientParams1.serial = "1"_sl;
+    caClientParams1.validity_secs = 3600*24;
+    Retained<Cert> cert1 = csr1->sign(caClientParams1, caKey, caCert);
+    cerr << "Cert1 info:\n" << string(cert1->summary("\t"));
+    
+    // Delete cert1 to cleanup:
+    Cert::deleteCert("cert1");
+    CHECK(Cert::loadCert("cert1") == nullptr);
+    
+    // Save cert1:
+    cert1->save("cert1", true);
+    Retained<Cert> cert1a = Cert::loadCert("cert1");
+    REQUIRE(cert1a);
+    CHECK(cert1a->data() == cert1->data());
+    
+    // Create CSR2:
+    Retained<PrivateKey> key2 = PrivateKey::generateTemporaryRSA(2048);
+    Retained<CertSigningRequest> csr2 = new CertSigningRequest(DistinguishedName(kSubject2Name), key2);
+    CHECK(csr2->subjectName() == kSubject2Name);
+    CHECK(csr2->subjectPublicKey()->data(KeyFormat::Raw) == key2->publicKey()->data(KeyFormat::Raw));
+    
+    // Sign and create cert2 with the same CA Cert as the cert1:
+    Cert::IssuerParameters caClientParams2;
+    caClientParams2.serial = "2"_sl;
+    caClientParams2.validity_secs = 3600*24;
+    Retained<Cert> cert2 = csr2->sign(caClientParams2, caKey, caCert);
+    cerr << "Cert2 info:\n" << string(cert2->summary("\t"));
+    
+    // Delete cert2 to cleanup:
+    Cert::deleteCert("cert2");
+    CHECK(Cert::loadCert("cert2") == nullptr);
+    
+    // Save cert2:
+    cert2->save("cert2", true);
+    Retained<Cert> cert2a = Cert::loadCert("cert2");
+    REQUIRE(cert2a);
+    CHECK(cert2a->data() == cert2->data());
+    
+    // Load cert1 again to make sure that it's still loaded:
```

Comment:
to understand more, is this some tricky way to validate chain of certs?

Reply:
Two cert chains are sharing the same intermediate cert so to make that saving the cert2 chain doesn't interfere with the cert1 chain.